crates/transaction-generator-lib/src/workflow_delegator.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    account_generator::AccountGeneratorCreator, accounts_pool_wrapper::AccountsPoolWrapperCreator,
+    call_custom_modules::CustomModulesDelegationGeneratorCreator,
+    entry_points::EntryPointTransactionGenerator, EntryPoints, ObjectPool,
+    ReliableTransactionSubmitter, TransactionGenerator, TransactionGeneratorCreator, WorkflowKind,
+};
+use aptos_logger::{info, sample, sample::SampleRate};
+use aptos_sdk::{
+    transaction_builder::TransactionFactory,
+    types::{transaction::SignedTransaction, LocalAccount},
+};
+use std::{
+    cmp,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+    time::Duration,
+};
+
+/// Wrapper that allows inner transaction generator to have unique accounts
+/// for all transactions (instead of having 5-20 transactions per account, as default)
+/// This is achieved via using accounts from the pool that account creatin can fill,
+/// and burning (removing accounts from the pool) them - basically using them only once.
+/// (we cannot use more as sequence number is not updated on failure)
+struct WorkflowTxnGenerator {
+    stage: StageTracking,
+    generators: Vec<Box<dyn TransactionGenerator>>,
+    pool_per_stage: Vec<Arc<ObjectPool<LocalAccount>>>,
+    num_for_first_stage: usize,
+    completed_for_first_stage: Arc<AtomicUsize>,
+}
+
+impl WorkflowTxnGenerator {
+    fn new(
+        stage: StageTracking,
+        generators: Vec<Box<dyn TransactionGenerator>>,
+        pool_per_stage: Vec<Arc<ObjectPool<LocalAccount>>>,
+        num_for_first_stage: usize,
+        completed_for_first_stage: Arc<AtomicUsize>,
+    ) -> Self {
+        Self {
+            stage,
+            generators,
+            pool_per_stage,
+            num_for_first_stage,
+            completed_for_first_stage,
+        }
+    }
+}
+
+impl TransactionGenerator for WorkflowTxnGenerator {
+    fn generate_transactions(
+        &mut self,
+        account: &LocalAccount,
+        mut num_to_create: usize,
+    ) -> Vec<SignedTransaction> {
+        assert_ne!(num_to_create, 0);
+        let mut stage = self.stage.load_current_stage();
+
+        match &self.stage {
```

Comment:
I didn't understand the difference between WhenDone and ExternallySet variants, and why they need to be different.

Reply:
when using this in the executor benchmark, we want to move when done. 

when using txn emitter to send workload to real network, it is easier to track things and look at the results, if we give fixed amount of time for each stage, and change the stage externally - so that emitter stats can be created per-stage.


maybe this can be simplified later